docs/design/otlp.md

```diff
@@ -0,0 +1,171 @@
+
+## 2. Overall Structure of `opentelemetry-rust` and OTLP
+```mermaid
+graph TD
+    A[Application code] --> B[opentelemetry-api]
+    B --> C[opentelemetry-sdk processors]
+    C --> D[opentelemetry-otlp exporters]
+    D -->|gRPC / HTTP| E[OTel Collector / Backend]
+    %% Auxiliary exporters
+    C --> F[custom exporters]
+```
+
+Important details:
+
+* `opentelemetry-sdk` owns batching, aggregation, lifecycle; **exporters handle serialization, transport, and may implement retry policy**.
+* Each signal (trace/metric/log) has its own exporter type but they share common builder + transport modules.
+
+---
+
+## 3. Crate Layout
+```text
+src/
+  lib.rs                 # Re-exports + feature flags + protocol enum
+  span.rs | metric.rs | logs.rs   # Signal-specific builders/exporters
+  exporter/
+     ├── mod.rs          # Common builder traits, env var parsing, compression
+     ├── http/           # Reqwest blocking/async clients, body encoder
+     └── tonic/          # Tonic clients, TLS/compression helpers
+```
+### 3.1 Common Interfaces
+* **`HasExportConfig` / `WithExportConfig`** – traits mixed into builders to expose shared config (`endpoint`, `timeout`, `protocol`).
+* **Builder marker types** (`NoExporterBuilderSet`, `HttpExporterBuilderSet`, `TonicExporterBuilderSet`) enforce at compile time that exactly *one* transport is chosen.
+* **`SupportedTransportClient` enum** – run-time dispatch inside the exporter when sending.
+
+---
+
+## 4. Feature-Flag Matrix
+| Cargo feature | Purpose | Conditional modules |
+|---------------|---------|---------------------|
+| `trace` / `metrics` / `logs` | Enable signal exporters | `span.rs`, `metric.rs`, `logs.rs` |
+| `grpc-tonic` | Use `tonic` gRPC transport | `exporter/tonic` |
+| `http-proto` *(default)* | HTTP + protobuf body | `exporter/http` |
+| `http-json` | HTTP + JSON body | `exporter/http` |
+| `gzip-tonic` `zstd-tonic` | gRPC message compression | `exporter/tonic` |
+| `reqwest-client` / `reqwest-blocking-client` *(default)* | Choose async vs blocking Reqwest HTTP client |
+| `hyper-client` | Use Hyper HTTP transport *(requires both Reqwest features disabled)* |
+| TLS helpers (`tls-roots`, `tls-webpki-roots`, `reqwest-rustls`) | Supply trust roots for TLS |
+
+Because **only one transport is valid per exporter**, `protocol` is currently a *hint* – unsupported values are ignored by the concrete builder.
+
+### 4.1 Transport-specific Client Options
+
+Exporters must function both in binaries that start no async runtime and in services already running Tokio.  Consequently the crate provides **blocking** and **non-blocking** HTTP clients behind mutually-exclusive feature-flags.  The blocking variant is the default to ensure out-of-the-box operability; applications that already depend on Tokio can opt into an async client to avoid spawning extra threads.
+
+| Implementation | Feature flag | Blocking? | Underlying crate | Transports | Typical use case |
```

Comment:
Unclear what we mean by "Blocking" - this could be quite misleading as users might think OTel blocks user-threads..

Reply:
fair; I'll use 'synchronous' which makes more sense in the context of the subsequent rows